### PR TITLE
Esp32WiFiManager Updates (SoftAP support)

### PR DESCRIPTION
**Pull request by atanisoft**
1) Adding SoftAP support using the provided ssid/password when running in AP only mode. When operating in AP+STA the SoftAP will use the hostname for it's SSID.
2) Use the full system_event_t object in process_wifi_event.
3) Add support for callers to ask the wifi mgr for event callbacks.
4) Added defines for silencing most of the Esp32SocketParams::log_message output.
5) Changed to using ESP_ERROR_CHECK_WITHOUT_ABORT to mdns_lookup rather than checking/logging the error in this method.
6) Added define to silence mdns_lookup output for search result output.
7) Shorten task name to "Esp32WiFiMgr".
8) Move mdns system init to STA start event handler.
9) Performance/stability improvements to the LwIP stack based on values coming in arduino-esp32 1.0.3.
10) #ifdef -> #if for feature flags
11) Add check for WDT being enabled before using APIs
12) Expose Notifiable::notify_from_isr based on OPENMRN_FEATURE_MUTEX_FREERTOS (underlying impl uses this flag)
13) Increase of LOG(...) buffer size to 2kb for ESP32
14) Replace a couple lingering references to HAVE_BSDSOCKET with OPENMRN_FEATURE_BSD_SOCKET.

### CODE REVIEW

File: src/executor/Notifiable.hxx [hunk 1]
```diff
@@ -46,7 +46,7 @@ class Notifiable : public Destructable
 public:
     /// Generic callback.
     virtual void notify() = 0;
-#ifdef __FreeRTOS__
+#if OPENMRN_FEATURE_MUTEX_FREERTOS
```
Comment (balazsracz) [inline src/executor/Notifiable.hxx]: I'm thinking whether this is the best feature macro to use here, vs. adding a new feature test macro.
Ultimately _from_isr() would need to exist for any RTOS, including an RTOS that has say pthreads APIs.

File: src/freertos_drivers/esp32/Esp32HardwareCanAdapter.hxx [hunk 2]
```diff
@@ -236,7 +240,7 @@ private:
             }
 
             /// ESP32 native CAN driver frame
-            can_message_t msg = {0};
+            can_message_t msg = {};
```
Comment (balazsracz) [inline src/freertos_drivers/esp32/Esp32HardwareCanAdapter.hxx]: what is the reason for this change? Will the new code also zero initialize?

File: src/os/OSSelectWakeup.cxx [hunk 3]
```diff
@@ -10,7 +10,7 @@ void empty_signal_handler(int)
 #include <sys/stat.h>
 #include <fcntl.h>
 
-#include "../vfs/esp_vfs.h"
+#include <esp_vfs.h>
```
Comment (balazsracz) [inline src/os/OSSelectWakeup.cxx]: are you sure this also compiles under the arduino UI (as opposed to native IDF and whatever build system you use).

File: src/utils/logging.cxx [hunk 4]
```diff
@@ -35,6 +35,8 @@
 
 #if defined(__linux__) || defined(__MACH__)
 char logbuffer[4096];
+#elif defined(ESP32)
+char logbuffer[2048];
```
Comment (balazsracz) [inline src/utils/logging.cxx]: 256 was not enough for your log lines?

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx [hunk 5]
```diff
@@ -372,93 +407,238 @@ void Esp32WiFiManager::factory_reset(int fd)
...
-            // connect process.
-            if (xEventGroupGetBits(wifiStatusEventGroup_) & WIFI_CONNECTED_BIT)
+        // Set the default mDNS instance name to the generated hostname.
+        ESP_ERROR_CHECK(mdns_instance_name_set(hostname_.c_str()));
+
+        if (stationStaticIP_)
+        {
+            // Stop the DHCP service before connecting, this allows us to
+            // specify a static IP address for the WiFi connection
+            LOG(INFO, "[DHCP] Stoping DHCP Client (if running).");
```
Comment (balazsracz) [inline src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: pp

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx [hunk 6]
```diff
@@ -372,93 +407,238 @@ void Esp32WiFiManager::factory_reset(int fd)
...
+        // Start the SSID connection process.
+        esp_wifi_connect();
+    }
+    else if (event->event_id == SYSTEM_EVENT_STA_CONNECTED)
+    {
+        LOG(INFO, "[WiFi] Connected to SSID: %s", ssid_);
+        // Set the flag that indictes we are connected to the SSID.
+        xEventGroupSetBits(wifiStatusEventGroup_, WIFI_CONNECTED_BIT);
+    }
+    else if (event->event_id == SYSTEM_EVENT_STA_GOT_IP)
```
Comment (balazsracz) [inline src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: is this invoked also when we are in static IP mode or in AP mode?

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx [hunk 7]
```diff
@@ -372,93 +407,238 @@ void Esp32WiFiManager::factory_reset(int fd)
...
+                                          softAPStaticIP_));
+
+            // Convert the Soft AP Static IP to a uint32 for manipulation
+            uint32_t apIP = ntohl(ip4_addr_get_u32(&softAPStaticIP_->ip));
+
+            // Default configuration is for DHCP addresses to follow
+            // immediately after the static ip address of the Soft AP.
+            ip4_addr_t first_ip, last_ip;
+            ip4_addr_set_u32(&first_ip, htonl(apIP + 1));
+            ip4_addr_set_u32(&last_ip, htonl(apIP + softAPMaxStations_));
```
Comment (balazsracz) [inline src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: reserve a few dozen IPs instead. If someone connects a wifi extender you get to deal with a lot more MAC addresses than 4.

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx [hunk 8]
```diff
@@ -507,109 +684,156 @@ void Esp32WiFiManager::start_wifi_system()
...
+
+    // If we are using the STATION interface, this will block until the ESP32
+    // starts the connection process, note it may not have an IP address
+    // immediately thus the need to check the connection result a few times
+    // before giving up with a FATAL error.
+    if (wifiMode_ == WIFI_MODE_APSTA || wifiMode_ == WIFI_MODE_STA)
+    {
+        uint8_t attempt = 0;
+        EventBits_t bits = 0;
+        uint32_t bitMask = WIFI_CONNECTED_BIT;
```
Comment (balazsracz) [inline src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: bit_mask

File: src/os/OSSelectWakeup.cxx (hunk 3 above)
Comment (atanisoft) [reply src/os/OSSelectWakeup.cxx]: I tested it on the 1.0.2 arduino-esp32 release and didn't see any failures. I also checked the default include path for 1.0.1 and newer versions and the vfs include directory is present

File: src/utils/logging.cxx (hunk 4 above)
Comment (atanisoft) [reply src/utils/logging.cxx]: For the CS it wasn't. Perhaps this could be smaller 1k instead of 2k though.

File: src/freertos_drivers/esp32/Esp32HardwareCanAdapter.hxx (hunk 2 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32HardwareCanAdapter.hxx]: Gcc complained on incomplete initializer. This is supposed to zero it as well but I have no problems putting in a memset instead of you prefer

File: src/executor/Notifiable.hxx (hunk 1 above)
Comment (atanisoft) [reply src/executor/Notifiable.hxx]: I wasn't sure either and considered adding a new define but all usages of this appeared to be on OSSem.notify_from_isr which used this define.

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 7 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: Esp32 only supports a max of 4 connected stations. I can increase it though.

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 5 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: Pp?

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx [hunk 9]
```diff
@@ -289,15 +324,15 @@ ConfigUpdateListener::UpdateAction Esp32WiFiManager::apply_configuration(
...
     if (lseek(fd, cfg_.offset(), SEEK_SET) != cfg_.offset())
     {
-        LOG(WARNING, "lseek failed to reset fd offset, REBOOT_NEEDED");
+        LOG_ERROR("lseek failed to reset fd offset, REBOOT_NEEDED");
         return ConfigUpdateListener::UpdateAction::REBOOT_NEEDED;
     }
 
     // If we are unable to read the full configuration from persistent storage
     // give up and request a reboot.
     if (read(fd, crcbuf.get(), cfg_.size()) != cfg_.size())
```
[the remark this answers is not in the document]
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: I'll look at the repeated_read and switch to that.

File: src/freertos_drivers/esp32/Esp32WiFiManager.hxx [hunk 10]
```diff
@@ -85,11 +87,42 @@ public:
...
+    /// @param hostname_prefix is the hostname prefix to use for this node.
+    /// The @ref NodeID will be appended to this value. The maximum length for
+    /// final hostname is 32 bytes.
+    /// @param wifi_mode is the WiFi operating mode. When set to WIFI_MODE_STA
+    /// the Esp32WiFiManager will attempt to connect to the provided WiFi SSID.
+    /// When the wifi_mode is WIFI_MODE_AP the Esp32WiFiManager will create an
+    /// AP with the provided SSID and PASSWORD. When the wifi_mode is
+    /// WIFI_MODE_APSTA the Esp32WiFiManager will connect to the provided WiFi
+    /// AP and create an AP with the SSID of "<hostname>" and the provided
+    /// password. Note, the password for the AP will not be used if
```
[the remark this answers is not in the document]
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.hxx]: I can add a pw for AP with default to the station pw when null and mode is not open

File: src/freertos_drivers/esp32/Esp32WiFiManager.hxx [hunk 11]
```diff
@@ -85,11 +87,42 @@ public:
...
     /// node uptime.
     Esp32WiFiManager(const char *ssid, const char *password,
-        openlcb::SimpleCanStack *stack, const WiFiConfiguration &cfg);
+        openlcb::SimpleCanStack *stack, const WiFiConfiguration &cfg,
+        const char *hostname_prefix = "esp32_",
+        wifi_mode_t wifi_mode = WIFI_MODE_STA,
+        tcpip_adapter_ip_info_t *station_static_ip = nullptr,
+        ip_addr_t primary_dns_server = ip_addr_any,
+        uint8_t soft_ap_channel = 1,
+        uint8_t soft_ap_max_stations = 4,
```
[the remark this answers is not in the document]
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.hxx]: Likely not, at least I can't think of anything specific. On the CS this is intended mainly as means for initial config but it can also be used for non-home (ie: train show) environments.

File: src/freertos_drivers/esp32/Esp32WiFiManager.hxx [hunk 12]
```diff
@@ -189,8 +236,9 @@ private:
     /// periodic health checks of the connected hubs or clients.
     os_thread_t wifiTaskHandle_;
 
-    /// Dynamically generated hostname for this node, esp32_{node-id}.
-    std::string hostname_{"esp32_"};
+    /// Dynamically generated hostname for this node, esp32_{node-id}. This is
+    /// also used for the SoftAP SSID name (if enabled).
+    std::string hostname_{""};
```
[the remark this answers is not in the document]
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.hxx]: Will drop, I shifted it to the constructor and forgot to clean this up

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 6 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: Yes, it is raised almost immediately after connect in that case. I can add it to the bit_mask when static IP is used so we don't block except for connect

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 5 above)
Comment (balazsracz) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: yes, you have a typo here.

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 6 above)
Comment (balazsracz) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: no that's fine. Less code is better. Just wanted to make sure it actually works :)

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 7 above)
Comment (balazsracz) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: The number of stations does not necessarily equal to the number of MAC addresses visible. If you add a wifi extender, you will see only one station (the wifi extender) but a bunch of MAC addresses that need separate IP addresses.

File: src/freertos_drivers/esp32/Esp32WiFiManager.hxx (hunk 10 above)
[the remark this answers is not in the document]
Comment (balazsracz) [reply src/freertos_drivers/esp32/Esp32WiFiManager.hxx]: sounds good

File: src/freertos_drivers/esp32/Esp32WiFiManager.hxx (hunk 11 above)
[the remark this answers is not in the document]
Comment (balazsracz) [reply src/freertos_drivers/esp32/Esp32WiFiManager.hxx]: i would recommend to remove this configuration option as it is unnecessary complexity in your code.

File: src/executor/Notifiable.hxx (hunk 1 above)
Comment (balazsracz) [reply src/executor/Notifiable.hxx]: let's add a new feature macro OPENMRN_FEATURE_NOTIFY_FROM_ISR

File: src/freertos_drivers/esp32/Esp32HardwareCanAdapter.hxx (hunk 2 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32HardwareCanAdapter.hxx]: switching to memset instead.

File: src/utils/logging.cxx (hunk 4 above)
Comment (atanisoft) [reply src/utils/logging.cxx]: dropped to 1k

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 7 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: got it, setting up a constant with a default value of 48 and capping at max of 4 stations (max for esp32).

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 8 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: fixed

File: src/freertos_drivers/esp32/Esp32WiFiManager.cxx (hunk 5 above)
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.cxx]: sorry missed that, fixed

File: src/freertos_drivers/esp32/Esp32WiFiManager.hxx (hunk 11 above)
[the remark this answers is not in the document]
Comment (atanisoft) [reply src/freertos_drivers/esp32/Esp32WiFiManager.hxx]: removed

File: src/executor/Notifiable.hxx (hunk 1 above)
Comment (atanisoft) [reply src/executor/Notifiable.hxx]: I'll update the refs that are in this PR to this, I spotted a few others that might make sense but will hold them back for a diff PR since I'm not certain of their usages.
